runtime/doc/api.txt

```diff
@@ -2563,7 +2563,8 @@ nvim_buf_get_extmarks({buffer}, {ns_id}, {start}, {end}, {opts})
 
     Parameters: ~
       • {buffer}  Buffer handle, or 0 for current buffer
-      • {ns_id}   Namespace id from |nvim_create_namespace()|
+      • {ns_id}   Namespace id from |nvim_create_namespace()| or -1 for all
+                  namespaces
```

Comment:
Could we have a value to just get anonymous extmarks? Would `0` work?

Reply:
Or maybe `-1` for anonymous namespace (which is consistent with `nvim_buf_add_highlight`), and `0` for all namespaces.

Reply:
I think so yes.

Reply:
We could go either way since `nvim_set_hl` uses `0` for un-namespaced highlights.

Reply:
Yes, that should be consistent: `0` for the "global" (anonymous) namespace, `-1` for all.

Reply:
But it's already inconsistent across existing APIs.

Reply:
No reason to add _more_ inconsistency ;)

Reply:
(and which other extmarks API uses `0` or `-1` inconsistently?)

Reply:
Do we even need `nvim_buf_add_highlight` (which has `FUNC_API_SINCE(1)`)? Seems like that's superseded by `nvim_buf_set_extmark`.

Reply:
> Well, looks like a fast path for `extmark_set`? 

I wouldn't say so, it's just less. I don't think anyone should be using it.

Reply:
Well, doing less is one of the obvious ways of having better performance...

Reply:
I doubt it in this case. If you pass empty opts, then very little in that function happens.

Reply:
Sure, as I said "I don't know". If I _wanted_ to know, I'd probably benchmark.

Reply:
> Yes, that should be consistent: 0 for the "global" (anonymous) namespace, -1 for all.

OK I'll change it to that. Although each anonymous namespace has it's own namespace id so it's not really global. I think the "global" term only makes sense in regards to highlights?

Reply:
`global` is just a term we use to describe highlights without a namespace.

Reply:
Indeed; ignore my suggestion then.

(Really, the proper thing to do would be to deprecate `nvim_buf_get_extmarks` and write a new one, with blackjack and `opts` -- which would then have fields for filtering.)

Reply:
I think the suggestion is still valid, just `0` for all anonymous namespaces, `-1` for all namespaces and `> 0` for a named namespace.

I don't see what deprecating would be necessary, the `opts` field is already there and the `ns_id` field can be repurposed as above.

Reply:
`nvim_buf_add_highlight` could just be deprecated already. `nvim_buf_add_highlight_electric_boogaloo` already exists, it is called `nvim_buf_set_extmark` (plus any language specific wrapper, as needed/wanted).

Reply:
> either you get all marks from a given ns_id as a number or you get marks from all namespaces

Yeah that was my idea as well. What was your usecase for special casing anonymous namespaces @lewis6991?

Reply:
Honestly I don't really have one. Just thought it would be "nice to have" and not really cost us anything since `0` is unused anyway.

Reply:
> Just thought it would be "nice to have"

Yeah I kinda agree, I had a commit ready implementing it but if we rather not introduce the concept of "all unnamed namespaces" then 🤷🏻

Reply:
One reason for adding a `0 -> anonymous namespaces` mapping is that one can already requests marks from a named namespace by inspecting `nvim_buf_get_namespaces()`. If you then know that you're not interested in the marks set by the named namespaces you might want to inspect the anonymous namespaces. If we only have `-1 -> all namespaces` you would then have to filter out all the marks belonging to a named namespace yourself.

Reply:
But yeah that is assuming that you first loop over `nvim_buf_get_namespaces()` which might be negated by simply supplying `-1` in the first place.

Reply:
It's very low cost to add and maintains some kind of consistency with `nvim_set_hl`.

I think your argument is true though. You might loop over `nvim_buf_get_namespaces` and only conditionally look at some namespaces as you might specifically exclude some (e.g. foldsigns.nvim excludes gitsigns extmarks), but then you still might want to look at the anonymous mark.

Reply:
> "anonymous namespaces" in plural sounds like a red herring. 

How come? The namespaces do exist, they just have no name and thus not returned by `nvim_get_namespaces`. It's useful to be able to segment these from marks in named namespaces.

> That is something which might be done using some stuffed away helper function.

How would one implement such a function? The API is supposed to be low-level. If it's not done there, where else can it be done? How would one handle the above usecase of needing to filter out specific namespaces? Simply getting all the marks isn't ideal.

Reply:
Yeah why do we have unnamed namespaces in the first place? But since we do, I think the `get_namespaces` API should have mapped from id -> name such that it would actually list all the namespaces.

Reply:
`nvim_get_namespaces2` ?

Reply:
Sorry, are you saying you'd like the `allow filtering nvim_buf_get_extmarks() by type` commit to be broken out to a separate PR?

Reply:
> Sorry, are you saying you'd like the allow filtering nvim_buf_get_extmarks() by type commit to be broken out to a separate PR?

github loading of comments seems to be broken, I always see more messages than the one I responded to after the fact... This was meant to be a direct response to the discussion about namespaces, not anything about types.

Reply:
> github loading of comments seems to be broken, I always see more messages than the one I responded to after the fact...

Yeah same here.

> This was meant to be a direct response to the discussion about namespaces, not anything about types.

> A general discussion about negative filtering is one that could be had (which that can filter out more meaningful patterns as well) . but I would rather like that be broken out to a separate issue. 

Alright I thought the above pertained to the filtering by extmark type, which I do think warrants some discussion/feedback. I.e. I only later realized that `number/line/cursorline_hl_group` are the counterparts of `:sign num/line/culhl`. I grouped them under `"highlight"` extmarks. I think the fact that these are part of the legacy sign API is irrelevant for extmarks?

Reply:
> :sign num/line/culhl. I grouped them under "highlight" extmarks. I think the fact that these are part of the legacy sign API is irrelevant for extmarks?

I agree with this reasoning, except for "num" as number column is not part of text. it is more like using the number column for a sign.

Reply:
> Getting all anonymous namespaces will already group together namespaces from unrelated plugins. Simply getting all namespaces in not a big jump from that conceptually. You will need to do similar work in post-processing to understand what is going on.

With the described usecase, that's exactly what's wanted. Named namespaces will have already been processed.
We don't want to then fetch them all again (creating potentially hundreds of Lua tables), especially when this code is done in on_line events.

> A general discussion about negative filtering is one that could be had (which that can filter out more meaningful patterns as well) . but I would rather like that be broken out to a separate issue. The singular anonymous namespace already handles the case for API completeness and compatibility with nvim_buf_add_highlight, which is all that is needed for this PR. (by itself a concept that should be deprecated, but it is a per-existing concept and thus needs to be supported)

Well such discussion could have an impact on the design of this API, so it is relevant to discuss here. Also, the PR also introduces a filter option which makes such discussion even more relevant.

I think we want something more general than negative filtering. We want access to all marks that isn't just "all or single ns".  It seems the best solution is replacing `nvim_get_namespaces` with something that actually returns the anonymous namespaces so one can do:

```lua
for ns, name in pairs(vim.api.nvim_get_namespaces2) do
  -- anonymous: name == ''
  ...
  for _, mark in ipairs(vim.api.nvim_get_extmarks(buf, ns, ...) do
    ...
  end
end  
```

Reply:
I agree that it is unfortunate that `nvim_get_namespaces` doesn't return the unnamed ones but what would `nvim_get_namespaces2` actually allow you to do/avoid to do that you can't do with a list of all extmarks (`ns_id == -1` here)? A namespace id -> name mapping is useless in and of itself without actually inspecting the marks within it right?

Reply:
Yeah I guess if it's a valid usecase to ignore on the basis of the namespace name.
In this usecase you are already looping over all namespaces so I would guess the actual content of the marks is what's interesting, not the namespace name.

Reply:
One specific usecase is for mirroring marks in other windows (for nvim-treesitter-context) and lines (foldsigns.nvim). Both of which want access to anonymous marks **and** want to exclude specific groups (e.g. Gitsigns and treesitter).

Reply:
> With the described usecase, that's exactly what's wanted. Named namespaces will have already been processed.
We don't want to then fetch them all again (creating potentially hundreds of Lua tables), especially when this code is done in on_line events.

Sure, the idea was we either fetch all namespaces at once (which sounded like what your plugin wanted, as you do after all process all marks, but doesn't work for some reason, which is fine), or all namespaces individually, even the unnamed ones in the latter case (which is I already agreed to nvim_get_namespaces is incomplete).

Currently all valid namespaces (except for the singular anonymous one) form a continuous range of id numbers. But I don't think we ever document this as a stable guarantee, so either we need to do so (strongly), or introduce this new function which returns all valid id:s.

Reply:
The plugin wants to get all the marks for a folded region with exclusions. Grabbing all the marks in one go will work, but it's also inefficient (mostly due to the Lua bridge). It's much more efficient to loop through the namespaces and only grab the marks you wish to process. E.g. we would absolutely want to exclude treesitter marks, there's no need to constantly keep pushing those through the Lua bridge.

Reply:
> Currently all valid namespaces (except for the singular anonymous one) form a continuous range of id numbers.

Right, then I guess a `ns_id -> name get_namespaces2()` could simply be replaced by a while loop:
```lua
local nsmap = {}
for k, v in pairs(vim.api.nvim_get_namespaces()) do
  nsmap[v] = k
end

local ns = 1
while true do
  if not nsmap[ns] then
    -- unnamed namespace
  end
  if not nsmap[ns]:find("ignorepattern") then
    -- ignore namespace by name
  end
  local err, marks = pcall(vim.api.nvim_buf_get_extmarks(buf, ns,...))
  if not err then
    -- ns > next_namespace_id
  end
  ns = ns + 1
end
```

Reply:
Not sure the above is more efficient than getting all marks at once and filtering those though. `next_namespace_id` can get quite large.